// vueuc/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VFocusTrap = exports.FocusTrap = exports.Overflow = exports.VOverflow = exports.VXScroll = exports.XScroll = exports.resizeObserverManager = exports.VResizeObserver = exports.ResizeObserver = exports.VLazyTeleport = exports.LazyTeleport = exports.VVirtualList = exports.VirtualList = void 0;
__exportStar(require("./binder/src"), exports);
var src_1 = require("./virtual-list/src");
Object.defineProperty(exports, "VirtualList", { enumerable: true, get: function () { return __importDefault(src_1).default; } });
Object.defineProperty(exports, "VVirtualList", { enumerable: true, get: function () { return __importDefault(src_1).default; } });
var src_2 = require("./lazy-teleport/src");
Object.defineProperty(exports, "LazyTeleport", { enumerable: true, get: function () { return __importDefault(src_2).default; } });
Object.defineProperty(exports, "VLazyTeleport", { enumerable: true, get: function () { return __importDefault(src_2).default; } });
var src_3 = require("./resize-observer/src");
Object.defineProperty(exports, "ResizeObserver", { enumerable: true, get: function () { return __importDefault(src_3).default; } });
Object.defineProperty(exports, "VResizeObserver", { enumerable: true, get: function () { return __importDefault(src_3).default; } });
Object.defineProperty(exports, "resizeObserverManager", { enumerable: true, get: function () { return src_3.resizeObserverManager; } });
var src_4 = require("./x-scroll/src");
Object.defineProperty(exports, "XScroll", { enumerable: true, get: function () { return __importDefault(src_4).default; } });
Object.defineProperty(exports, "VXScroll", { enumerable: true, get: function () { return __importDefault(src_4).default; } });
var overflow_1 = require("./overflow");
Object.defineProperty(exports, "VOverflow", { enumerable: true, get: function () { return overflow_1.VOverflow; } });
Object.defineProperty(exports, "Overflow", { enumerable: true, get: function () { return overflow_1.Overflow; } });
var focus_trap_1 = require("./focus-trap");
Object.defineProperty(exports, "FocusTrap", { enumerable: true, get: function () { return focus_trap_1.FocusTrap; } });
Object.defineProperty(exports, "VFocusTrap", { enumerable: true, get: function () { return focus_trap_1.FocusTrap; } });
