// @vuepress/plugin-palette/lib/node/palettePlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.palettePlugin = void 0;
const chokidar = require("chokidar");
const preparePaletteFile_1 = require("./preparePaletteFile");
const prepareStyleFile_1 = require("./prepareStyleFile");
const presetOptions_1 = require("./presetOptions");
const palettePlugin = ({ preset = 'css', userPaletteFile = presetOptions_1.presetOptions[preset].userPaletteFile, tempPaletteFile = presetOptions_1.presetOptions[preset].tempPaletteFile, userStyleFile = presetOptions_1.presetOptions[preset].userStyleFile, tempStyleFile = presetOptions_1.presetOptions[preset].tempStyleFile, importCode = presetOptions_1.presetOptions[preset].importCode, } = {}) => ({
    name: '@vuepress/plugin-palette',
    alias: (app) => ({
        '@vuepress/plugin-palette/palette': app.dir.temp(tempPaletteFile),
        '@vuepress/plugin-palette/style': app.dir.temp(tempStyleFile),
    }),
    onPrepared: async (app) => {
        await Promise.all([
            (0, preparePaletteFile_1.preparePaletteFile)(app, {
                userPaletteFile,
                tempPaletteFile,
                importCode,
            }),
            (0, prepareStyleFile_1.prepareStyleFile)(app, {
                userStyleFile,
                tempStyleFile,
                importCode,
            }),
        ]);
    },
    onWatched: (app, watchers) => {
        const paletteWatcher = chokidar.watch(userPaletteFile, {
            cwd: app.dir.source(),
            ignoreInitial: true,
        });
        paletteWatcher.on('add', () => {
            (0, preparePaletteFile_1.preparePaletteFile)(app, {
                userPaletteFile,
                tempPaletteFile,
                importCode,
            });
        });
        paletteWatcher.on('unlink', () => {
            (0, preparePaletteFile_1.preparePaletteFile)(app, {
                userPaletteFile,
                tempPaletteFile,
                importCode,
            });
        });
        watchers.push(paletteWatcher);
        const styleWatcher = chokidar.watch(userStyleFile, {
            cwd: app.dir.source(),
            ignoreInitial: true,
        });
        styleWatcher.on('add', () => {
            (0, prepareStyleFile_1.prepareStyleFile)(app, {
                userStyleFile,
                tempStyleFile,
                importCode,
            });
        });
        styleWatcher.on('unlink', () => {
            (0, prepareStyleFile_1.prepareStyleFile)(app, {
                userStyleFile,
                tempStyleFile,
                importCode,
            });
        });
        watchers.push(styleWatcher);
    },
});
exports.palettePlugin = palettePlugin;
